### ISSUE TITLE
Suggestion: provide a option prevent Duplicates on same type toast

### ISSUE THREAD

**Issue by bung87** [Maronato/vue-toastification#13]
like https://github.com/artemsky/vue-snotify preventDuplicates do, It prevent basic type like info, error etc.
I think it's better let developer using custom type(category, group) when calling $toast.
the background story is I want use it as http request fails callback globally, so  there may have many times fails same time.

**Comment by trevorgehman** [comment]
I have a similar suggestion. Here are a couple thoughts:

- Let the `maxToasts` be different for each position. (So I can limit it to 1 for the bottom-center, and 10 for the top-right, for example.)
- Let us have a `maxToastsType` option so we can either queue or discard additional notifications of a certain type. Perhaps allow this to be set either globally or on a per-position basis.
- Let us create our own types. This way we aren't limited to just using the predefined types.

**Comment by Maronato** [comment]
Hey @bung87! Sorry about the delay.
The use case does indeed seem valid, but I have some questions regarding how the toasts should be displayed when multiple of the same type are requested. Should the toasts be "queued" like they are when setting `maxToasts` or should new toasts matching existing types just be ignored?

I took a look at [vue-snotify's implementation](https://github.com/artemsky/vue-snotify/blob/2e0412c283019c66c8202b8d3b353347b08e5b69/src/SnotifyService.ts#L90) and it looks like they just ignore new toasts which types are already being rendered.

@trevorgehman, the above question also applies to your `maxToastsType` suggestion. What should the behavior be?

 Also, @trevorgehman, what would be the benefits of creating your own types? Currently, the only difference between types is styling. In fact, using `$toast("my toast", { type: "custom" })` should work fine, apart from some type validation warnings throw by Vue and the fact that there won't be any styling applied to the toast.
Assuming the validation warnings were not thrown, the behavior would be the same as editing the styles of current toasts.

**Comment by trevorgehman** [comment]
> @trevorgehman, the above question also applies to your `maxToastsType` suggestion. What should the behavior be?

I think queuing is preferable to dropping. But probably it should be an option with a default, like `queueAdditionalToasts:true`. For instance, if you want to show a notification for a network error, it would be good to drop duplicates so you don't get 10 "Network error" notifications in a row.

> Also, @trevorgehman, what would be the benefits of creating your own types? Currently, the only difference between types is styling. In fact, using `$toast("my toast", { type: "custom" })` should work fine, apart from some type validation warnings throw by Vue and the fact that there won't be any styling applied to the toast.

I like the idea of creating custom, very unique notifications using single file components. In that case, I would prefer not having any styling at all. I think having the flexibility to have any type you want would allow us to set separate `maxToasts` per position and per type. For instance, I might want to have a Network Error notification only show once on the top right (and discard any additional). But I want other notifications, like success messages, to show on the top right with a higher limit. What's the best way to accomplish that?

**Comment by bung87** [comment]
yeah,the keyword "queued" ,I'd like have an option to specify one type(category,or called group) has a limit, for example ,view initial may have multiple http request, when show "server error" I only want it show once at same time .

combine with @trevorgehman comments, the options like 
```
options = {
    "max": [
    	{
            "type":"all"
            "max": 2,
            "position":"all"
        },
         {
            "type":"error"
            "max": 1,
            "position":"bottom"
        },

    ]
}
Vue.use(Toast, options);
```

**Comment by Maronato** [comment]
I get the idea, but the usage seems to be getting a little too niche.
What if instead of implementing all of these options, there was a `filterToasts` callback that would receive the list of toasts to render and the user decides which logic to use to filter them?

I'd add your suggestions as code snippets, but you'd be able to customize it to your needs.

**Comment by trevorgehman** [comment]
> I get the idea, but the usage seems to be getting a little too niche.
> What if instead of implementing all of these options, there was a `filterToasts` callback that would receive the list of toasts to render and the user decides which logic to use to filter them?

Would we have the ability with the callback to render, queue or discard?

**Comment by bung87** [comment]
cool, think we can have them both, let filterToasts function use depends developer own, when it's value as string , we can provider preset filter function, so when new toast coming in ,the function context will have the incoming toast info ,and queued toasts info ,let developer decide what to do.

**Comment by trevorgehman** [comment]
Here's another consideration. Google has toast notifications inside Gmail. They limit it to showing one at a time. But instead of queuing or discarding, new toast notifications replace the previous one instantly:

![2019-11-07_00-11-43](https://user-images.githubusercontent.com/3209864/68361845-4d2bf200-00f3-11ea-890e-b3a8c1daf920.gif)

So as far as I can tell, in addition to setting a `maxToasts` setting, it would be nice to be able to determine what to do when the limit is reached:

- Discard
- Queue
- Replace

**Comment by Maronato** [comment]
Hey, @trevorgehman @bung87 I've implemented a solution that will be merged soon.
It works by exposing two callbacks that make it easy to arbitrarily filter toasts and control both enqueueing and discarding toasts.
I've also added code snippets to the documentation that implement the use cases that you guys mentioned.

Also, @trevorgehman, regarding replacing existing toasts, you can use `$toast.update` for that. One way would be to manually set the ID as the type (e.g. `success` or `error`) and always update that. I'll make a change to `.update` so that when it's called on a non-existing ID it creates it.

**Comment by Maronato** [comment]
Please see if v0.5.0 addresses your use cases. I'll reopen this issue if it does not.

**Comment by trevorgehman** [comment]
Yes @Maronato I think with with "createOrUpdate" functionality now available with `$toast.update` I can have a single notification that is instantly replaced if a new one comes in. So I'm good with the new version!

Also, just wanted to post another example, Dropbox, that does the same thing with their bottom notifications. This is a really common pattern from what I can tell:

![2019-11-11_15-59-48](https://user-images.githubusercontent.com/3209864/68621844-bb2f3b00-049e-11ea-8ff8-cda3dcd3e730.gif)

**Comment by trevorgehman** [comment]
Actually, one issue is that when I update an existing toast, it doesn't reset the timeout. I'd need some way to reset the timeout when updating an existing toast. It seems like if we pass a new `timeout` value in the options maybe the timeout should be reset in that case?

**Comment by Maronato** [comment]
It wouldn't be reset because the timeout is handled with a CSS animation, not with JS. I've fixed it in the newest version, 0.5.1, which I just released.

By default, it won't reset the animation, but if you pass a new `timeout` value, or even the same one, it will as expected.
